ssh-server/main.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"os/exec"
+	"os/signal"
+	"syscall"
+
+	"github.com/kr/pty"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/crypto/ssh/terminal"
+)
+
+var (
+	serverPort = flag.Uint("port", uint(22), "the port to listen on")
+)
+
+func main() {
+	flag.Parse()
+
+	config, err := setupAuthentication()
+	if err != nil {
+		log.Fatalf("setup authentication: %v", err)
+	}
+
+	serverAddress := fmt.Sprintf(":%d", *serverPort)
+	listener, err := net.Listen("tcp", serverAddress)
+	if err != nil {
+		log.Fatalf("listen for connection: %v", err)
+	}
+	log.Printf("Listening on %s...", serverAddress)
+
+	for {
+		connection, err := listener.Accept()
+		if err != nil {
+			log.Fatalf("accept incoming connection: %v", err)
+		}
+
+		go handleConnection(connection, config)
+	}
+}
+
+func setupAuthentication() (*ssh.ServerConfig, error) {
+	config := &ssh.ServerConfig{
+		NoClientAuth: true,
+	}
+
+	privateBytes, err := ioutil.ReadFile("id_rsa")
+	if err != nil {
+		return nil, fmt.Errorf("read private server key: %v", err)
+	}
+	private, err := ssh.ParsePrivateKey(privateBytes)
+	if err != nil {
+		return nil, fmt.Errorf("parse private server key: %v", err)
+	}
+	config.AddHostKey(private)
+	return config, nil
+}
+
+func handleConnection(nConn net.Conn, serverConfig *ssh.ServerConfig) error {
+	// Before use, a handshake must be performed on the incoming
+	// net.Conn.
+	_, chans, reqs, err := ssh.NewServerConn(nConn, serverConfig)
+	if err != nil {
+		return fmt.Errorf("handshake: %v", err)
+	}
+
+	// Service the incoming request channel so connection doesn't hang
+	go ssh.DiscardRequests(reqs)
+
+	// Service the incoming channels
+	for newChannel := range chans {
+		if err = serviceChannel(newChannel); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func serviceChannel(newChannel ssh.NewChannel) error {
+	if newChannel.ChannelType() != "session" {
+		newChannel.Reject(ssh.UnknownChannelType, "unknown channel type")
+		return nil
+	}
+	channel, requests, err := newChannel.Accept()
+	if err != nil {
+		return fmt.Errorf("accept channel: %v", err)
+	}
+	defer channel.Close()
+
+	// Tell the client we accept pty and shell commands
+	go func(in <-chan *ssh.Request) {
+		for req := range in {
+			req.Reply(req.Type == "pty-req" || req.Type == "shell", nil)
+		}
+	}(requests)
+
+	// Start the command with a pseudo-terminal.
```

Comment:
This comment, too, doesn't add anything beyond what the code does.  In general, comments should describe _why_ a thing is happening, not _what_ (since the code does that).   For example, if a comment was needed here, it might be:
// Allocate a pty instead of using simple I/O redirection so terminal features like job control and resizing are available.